Add db migration worker deployment

roschaefer opened the pull request:
- [x] persist database data - if the neo4j pod gets deleted, it does not take the data with it
- [x] test and document data migration routine
- [x] configure backend to serve public folder via https (for uploads)
- [x] DRY and parameterize conguration

roschaefer committed:
Add deployment for db-migation-worker

roschaefer committed:
Remove obsolete deployment files

roschaefer committed:
Create persistent volumes for minikube

Running the import works in minikube!

roschaefer committed:
Enable apoc to run neo4j import

roschaefer committed:
Point the backend to the local neo4j service

roschaefer committed:
Use folders for groups of YAML files

This keeps our configuration DRY and helps us to save keystrokes.

roschaefer committed:
Improve README

roschaefer committed:
Point web service to backend of internal network

roschaefer committed:
Backend is accessible only from internal network

I think this is better. For development it might be interesting to have
access to the backend but I think one should do development with
docker-compose not minikube and such.

roschaefer committed:
Kubernetes best practices:

* put many configurations in one file
* expose services before you start pods

See: https://kubernetes.io/docs/concepts/configuration/overview/

roschaefer committed:
Base64 encode default secrets in template

The JWT_SECRET is taken from the current default secret in
`Nitro-Backend`, the MONGODB_PASSWORD is just the label encoded.

roschaefer committed:
Add commits to configuration to deploy :latest tag

The recommended way to update a kubernetes deployment to a new image is
to change the image tag. However, our build server is configured to push
every commit of the `master` branch to docker hub to the respective
repository and tag `:latest`. So adding some configuration that can be
changed seems to be a trick to re-deploy the `:latest` image.

See here:
https://stackoverflow.com/a/51835397

roschaefer committed:
Reproduced the whole setup locally with success

roschaefer commented:
@mattwr18 I just decided to do the rest in another PR. I would like get this merged as a first step :smile_cat:

roschaefer committed:
Fix certain configuration for Digital Ocean

appinteractive commented on README.md:
```diff
@@ -4,60 +4,96 @@
-# change all vars as needed and deploy it afterwards
-kubectl apply -f secrets.yaml
+# example how to base64 a string:
+$ echo -n 'admin' | base64
```
Why?

roschaefer replied:
Secrets are expected to be base64 encoded and get decoded. If you add a non-base64 encoded string, you get jibberish output for the environment variable in the pod

appinteractive commented on config/backend.yml:
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 data:
   GRAPHQL_PORT: "4000"
-  GRAPHQL_URI: "https://api-nitro-staging.human-connection.org"
+  GRAPHQL_URI: "http://nitro-backend.staging:4000"
```
That will break on staging 🤔

appinteractive commented on config/neo4j.yml:
```diff
@@ -1,7 +1,7 @@
 kind: ConfigMap
 data:
-  NEO4J_URI: "bolt://neo4j:7687"
+  NEO4J_URI: "bolt://neo4j.staging:7687"
```
This also would break I think

roschaefer replied:
No, that's default kubernetes ~behaviour~ magic. You can check yourself by
```
kubectl --namespace=staging exec -it <POD_ID> nslookup <SERVICE_NAME>.staging
```
The trailing `staging` is optional, if you omit it, it will match all services in all namespaces, as far as I know.

See here: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#a-records

appinteractive commented on deployments/neo4j.yml:
```diff
@@ -0,0 +1,92 @@
+      metadata:
+        labels:
+          workload.user.cattle.io/workloadselector: deployment-staging-neo4j
+        name: nitro-neo4j
```
This is the name that is used for connecting to the db internally via the local network `bolt://nitro-neo4j:7687`

roschaefer replied:
Exactly, that was the problem. :wink: :+1:

roschaefer committed:
Bundle all configuration in one folder staging/

roschaefer committed:
Adding dashboard deployment for digital ocean

roschaefer committed:
Put many configuration files in one folder

roschaefer committed:
Rename .yml to .yaml files for consistency

See:
https://stackoverflow.com/a/21059164

@appinteractive

roschaefer committed:
Remove obsolete volume in deployment

roschaefer committed:
Merge pull request #5 from Human-Connection/digital_ocean

Fix certain configuration for Digital Ocean

roschaefer commented:
@appinteractive can this be merged?

appinteractive commented:
>@appinteractive can this be merged?

@roschaefer your checkbox for serving assets is not checked. Just missed to update or is it in another pr?
Beside that It looks good to me.

appinteractive commented:
Only one thing would be good, to add the secrets.yaml to the global ignore file instead only the staging one. Just to be save.

roschaefer committed:
Refactor db-migration-worker deployment

Instead of creating a deployment with a replica set, we just create the
pod once. Also the pod should have everything in the future to run the
database migration. Ie. have `cypher-shell` to write directly to the
database in the current network. All required configuration is passed
manually to the `db-migration-worker`-pod directly. SSH-keys are copied
through a secrets file. This altogether made many configuration files
obsolete.

roschaefer committed:
Create a persistent volume claim for uploads

roschaefer committed:
Multiple container per pod setup

roschaefer committed:
Set NEO4J_URI to localhost

Apparently, if you have a mult-container setup, the same pod is
accessible via localhost.

roschaefer committed:
Update README

roschaefer committed:
Fix README

roschaefer committed:
Cleanly separate legacy migration part

Use `kubectl patch -f ...`

roschaefer commented:
@appinteractive I'm done, can we merge this PR?
